amazon/ion/core.py

```diff
@@ -375,27 +392,52 @@ class Timestamp(datetime):
     Notes:
         The ``precision`` field is passed as a keyword argument of the same name.
     """
-    __slots__ = [TIMESTAMP_PRECISION_FIELD, TIMESTAMP_FRACTION_PRECISION_FIELD]
+    __slots__ = [TIMESTAMP_PRECISION_FIELD, TIMESTAMP_FRACTION_PRECISION_FIELD, TIMESTAMP_FRACTIONAL_SECONDS_FIELD]
 
     def __new__(cls, *args, **kwargs):
         precision = None
         fractional_precision = None
+        fractional_seconds = None
         if TIMESTAMP_PRECISION_FIELD in kwargs:
             precision = kwargs.get(TIMESTAMP_PRECISION_FIELD)
             # Make sure we mask this before we construct the datetime.
             del kwargs[TIMESTAMP_PRECISION_FIELD]
         if TIMESTAMP_FRACTION_PRECISION_FIELD in kwargs:
             fractional_precision = kwargs.get(TIMESTAMP_FRACTION_PRECISION_FIELD)
-            if fractional_precision is not None and 1 > fractional_precision > MICROSECOND_PRECISION:
+            if fractional_precision is not None and fractional_precision < 1:
                 raise ValueError('Cannot construct a Timestamp with fractional precision of %d digits, '
-                                 'which is out of the supported range of [1, %d].'
-                                 % (fractional_precision, MICROSECOND_PRECISION,))
+                                 'which is less than 1.' % fractional_precision)
             # Make sure we mask this before we construct the datetime.
             del kwargs[TIMESTAMP_FRACTION_PRECISION_FIELD]
+        if TIMESTAMP_FRACTIONAL_SECONDS_FIELD in kwargs:
+            fractional_seconds = kwargs.get(TIMESTAMP_FRACTIONAL_SECONDS_FIELD)
+            # Make sure we mask this before we construct the datetime.
+            del kwargs[TIMESTAMP_FRACTIONAL_SECONDS_FIELD]
+
+        if args[6] is not None and fractional_precision is not None and fractional_seconds is not None:
```

Comment:
Please name `6` something like `MICROSECOND_ARGUMENT_INDEX`. Might as well also assign the value of `args[MICROSECOND_ARGUMENT_INDEX]` to a variable since it's used several times unconditionally.

Can the value of `microsecond` also occur only in `kwargs`, or is it always present at index 6 in `args`? What if all of the default arguments are used, or if microseconds is passed in using a named parameter?

Also, it looks like we'll need to set `fractional_precision` before this if `microsecond` is set but `fractional_precision` isn't. Or should we raise an error if one is present and the other isn't?